src/frequenz/sdk/timeseries/logical_meter/logical_meter.py

```diff
@@ -0,0 +1,111 @@
+"""A logical meter for calculating high level metrics for a microgrid.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+
+import asyncio
+import uuid
+from typing import Dict, List
+
+from frequenz.channels import Broadcast, Receiver, Sender
+
+from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...microgrid.component import ComponentMetricId
+from ...timeseries import Sample
+from ._formula_engine import FormulaBuilder, FormulaEngine
+
+
+class LogicalMeter:
+    """A logical meter for calculating high level metrics for a microgrid.
+
+    LogicalMeter instances need to be specified formulas for each of the metrics that
+    need to be calculated.
+
+    Formulas can have Component IDs that are preceeded by a pound symbol("#"), and these
+    operators: +, -, *, /, (, ).
+
+    For example, the input string: "#20 + #5" is a formula for adding metrics from two
+    components with ids 20 and 5.
+    """
+
+    def __init__(
+        self,
+        channel_registry: ChannelRegistry,
+        resampler_subscription_sender: Sender[ComponentMetricRequest],
+    ) -> None:
+        """Create a `LogicalMeter instance`.
+
+        Args:
+            channel_registry: A channel registry instance shared with the resampling
+                actor.
+            resampler_subscription_sender: A sender for sending metric requests to the
+                resampling actor.
+        """
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._namespace = f"logical-meter-{uuid.uuid4()}"
+
+        # Channels are created inside the class under the assumption that LogicalMeter
+        # won't be an actor.  If it is going to be an actor, then it can't own the
+        # channels anymore.
+        self._output_channels: Dict[str, Broadcast[Sample]] = {}
+        self._tasks: List[asyncio.Task[None]] = []
+
+    async def _engine_from_formula_string(
+        self, formula: str, metric_id: ComponentMetricId
+    ) -> FormulaEngine:
+        builder = FormulaBuilder(
+            self._namespace,
+            self._channel_registry,
+            self._resampler_subscription_sender,
+            metric_id,
+        )
+        return await builder.from_string(formula)
+
+    async def _run_formula(
+        self, formula: FormulaEngine, sender: Sender[Sample]
+    ) -> None:
+        """Run the formula repeatedly and send the results to a channel.
+
+        Args:
+            formula: The formula to run.
+            sender: A sender for sending the formula results to.
+        """
+        while msg := await formula.apply():
+            await sender.send(msg)
+
+    async def start_formula(
```

Comment:
Somehow I feel that the formula in the function name is not needed and `start` alone would be enough.

Reply:
I'm not sure.  This would be the method that actors would use to start formulas,  and it would just start formulas.  So it would look like: `logical_meter.start_formula(...)`.  If we have `logical_meter.start(...)`,  that looks like we're trying to start the logical meter,  and not just one formula.

Reply:
Makes sense. I haven't looked at it from that perspective ;).